### PR TITLE
Make astro package play nice with node16 module resolution

### PR DESCRIPTION
**Pull request by Alxandr**
Projects using node16 module resolution in typescript uses the new
exports and imports configuration from typescript to find definition
files. This mirrors how nodejs resolves the files. If a package contains
an exports map in the package.json, typescript will ignore the "types"
field (not sure how it plays with typesVersions). This moves the typings
hirearchy of definition files into the same hierarchy that astro
produces output files in, so that typescript can discover them.

## Changes

Projects using node16 module resolution in typescript uses the new
exports and imports configuration from typescript to find definition
files. This mirrors how nodejs resolves the files. If a package contains
an exports map in the package.json, typescript will ignore the "types"
field (not sure how it plays with typesVersions). This moves the typings
hirearchy of definition files into the same hierarchy that astro
produces output files in, so that typescript can discover them.

## Testing

I've used `pnpm link` locally to do `import type { AstroIntegration } from 'astro';` - it does not work if I just use the `astro` package from NPM.

## Docs

This should only be visible in that people who use the new module resolution can now get typings - and also, the file hierarchy of the dist folder in the package is modified in that type-declarations now live next to their js counterparts.

Fixes: #4172

### CODE REVIEW

**Comment by FredKSchott** [comment]
This all makes sense to me! Thanks for the investigation and the thorough PR. 

As Nate said in the issue I’d also want a +1 from out resident TS expert @Princesseuh before merging.

File: packages/astro/package.json
Comment (bluwy) [inline packages/astro/package.json:27]: Note: the `types` condition should come before `default`. ([typescript docs](https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing))

Comment (Alxandr) [reply packages/astro/package.json:27]: This is JSON, the order is arbitrary. I see the example in the docs has it before, but that's just an example, and nowhere in the text I read that it should come before (not as far as I can understand at least) - if I'm reading this wrong, please point out where I'm misreading.

Also do note that this is actually working, as I'm running locally with a `pnpm patch` to make astro work for me.

Comment (bluwy) [reply packages/astro/package.json:27]: It's shown in this code example in the linked docs:

image

and in the [announcement post](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#package-json-exports-imports-and-self-referencing):

image

Disclaimer: I've created [publint](https://publint.bjornlu.com), so I'm picky with this 🙂

Comment (Princesseuh) [reply packages/astro/package.json:27]: It doesn't really matter in this case, but Node's documentation also mention that `default` should always be last

Comment (Alxandr) [reply packages/astro/package.json:27]: I love being picky as long as there's clear rules, and in this case there seems to be. I think it's a bad idea to have order-dependent JSON, so I'd argue node made a bad decision here, but it's what we got to live with, I guess. I'll update the pr when I'm back at my computer.

Comment (natemoo-re) [reply packages/astro/package.json:27]: Oh wow, I didn't know about these rules! That's super annoying, but yeah it would be great to make sure we're following them exactly.

**Review by Princesseuh — CHANGES_REQUESTED** [review]
Thank you for making this PR, support for `moduleResolution: 'node16'` is definitely something we want in Astro! In addition to the comment about the order of the types, the path to the types should be updated in [`astro-jsx.d.ts`](https://github.com/withastro/astro/blob/main/packages/astro/astro-jsx.d.ts).

**Additionally, an important note about this:** While this unblock usage inside `.ts(x)` files, `moduleResolution: 'node16'` is currently unsupported in our [language server](https://github.com/withastro/language-tools/tree/main/packages/language-server) so inside `.astro` file, even if you set it to `node16`, `node` will be instead forced on. 

Hopefully this is not a dealbreaker for you to use Astro, we're definitely planning to support `node16` as soon as possible, but we're currently focused on other tasks.

Apart from that, it looks good to me. It's a bit hard to tell on such a large change, but I don't see any reasons it would break anything. However, just to be sure we don't break everything right before the release of 1.0 (tomorrow!), we'll wait after the release to merge this in, hope that's okay.

**Comment by Alxandr** [comment]
I'm working around the issue of astro-files not caring about `node16` by moving code to `.tsm` files and just importing them in astro-files - and for the most part that works (though it's definitely been less than optimal). I'm assuming the greater ecosystem will eventually migrate most of the things to work with node16, but it will definitely take time.

File: packages/astro/package.json
Comment (Alxandr) [reply packages/astro/package.json]: I have updated the order.

**Review by Princesseuh — APPROVED** [review]

**Review by matthewp — CHANGES_REQUESTED** [review]
Marking this a blocked for now as we're still discussing whether this should be a minor or patch release.

**Comment by matthewp** [comment]
This is marked as being a minor release, so it can go out the next time we do one of those.

**Review by matthewp — APPROVED** [review]

**Pull request merged**
